html/modules/custom/docstore/src/MetadataTrait.php

```diff
@@ -273,66 +272,251 @@ protected function findTargetByProperty($reference, $target, $field, $value) {
-    ];
-
-    // Set creation time.
-    $item['created'][] = [
-      'value' => time(),
-    ];
-
-    // Set owner.
-    $item['provider_uuid'][] = [
-      'target_uuid' => $provider->uuid(),
-    ];
-
-    // Check for meta tags.
-    if (isset($params['metadata']) && $params['metadata']) {
+  public function updateEntityFieldsFromParameters(FieldableEntityInterface $entity, array $params, array $protected_fields = []) {
+    $updated_fields = [];
+
+    // Update all fields specified in metadata.
+    if (isset($params['metadata'])) {
       $metadata = $params['metadata'];
-      $item = array_merge($item, $this->buildItemDataFromMetaData($metadata, $vocabulary->id(), $provider, $params['author'], 'term'));
+      if (!is_array($metadata) || $this->arrayIsAssociative($metadata)) {
+        throw new BadRequestHttpException('Metadata has to be an array');
+      }
+
+      foreach ($metadata as $metaitem) {
+        foreach ($metaitem as $name => $values) {
+          $updated_fields[$this->setEntityFieldValue($entity, $name, $values, $protected_fields)] = TRUE;
+        }
+      }
+      unset($params['metadata']);
     }
 
-    // Create term.
-    $term = Term::create($item);
+    // Update all fields specified in params.
+    foreach ($params as $name => $values) {
+      // Ignore revision fields.
+      if ($name === 'new_revision' || $name === 'revision_log' || $name === 'draft') {
+        continue;
+      }
+
+      $updated_fields[$this->setEntityFieldValue($entity, $name, $values, $protected_fields)] = TRUE;
+    }
 
-    // Check for invalid fields.
-    foreach ($item as $key => $data) {
-      if (!$term->hasField($key)) {
-        throw new \Exception(strtr('Unknown field @field', [
-          '@field' => $key,
-        ]));
+    return $updated_fields;
+  }
+
+  /**
+   * Set the field values of an entity.
+   *
+   * @param \Drupal\Core\Entity\FieldableEntityInterface $entity
+   *   Fieldable entity (ex: node, term).
+   * @param string $name
+   *   Field name.
+   * @param mixed $values
+   *   Field values.
+   * @param array $protected_fields
+   *   List of fields that cannot be changed.
+   *
+   * @return string
+   *   Field name.
+   *
+   * @throws \Symfony\Component\HttpKernel\Exception\BadRequestHttpException
+   *   400 Bad Request if the field cannot be changed or doesn't exist.
+   */
+  public function setEntityFieldValue(FieldableEntityInterface $entity, $name, $values, array $protected_fields = []) {
+    // Make sure protected fields aren't set.
+    if (isset($protected_fields[$name])) {
+      throw new BadRequestHttpException(strtr('Field @name cannot be changed', ['@name' => $name]));
+    }
+
+    if ($entity->hasField($name)) {
+      // @todo explain why we do that.
+      // @see https://github.com/UN-OCHA/docstore-site/pull/75
+      if (is_array($values) && !$this->arrayIsAssociative($values)) {
+        foreach ($values as $key => $value) {
```

Comment:
@attiks this is an adaptation of your modifications from https://github.com/UN-OCHA/docstore-site/pull/75. Could you explain what is for?

Reply:
Data is in 3 possible formats, there's probably a cleaner way to handle this

```json
{
  "key": "value"
}
```

```json
{
  "key": [
    "value",
    "value"
  ]
}
```

```json
{
  "key": {
    "property": "value"
  }
}
```

Reply:
@attiks Thanks. The code from https://github.com/UN-OCHA/docstore-site/pull/75 and here focuses on the `uuid`. Would you have examples of those different formats? For example from assessment registry? Is it something related to the external entities?

Reply:
- `"term_label": "string"`
- `"term_label": ["string1", "string2"]` or `"term": ["uuid", "uuid"]`
- `"geofield": {"lat": 1, "lon": 2}`

> Is it something related to the external entities?

No